From 299a54a55f8d296a29937d58b5b7a258bfbe861a Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Tue, 26 May 2020 15:04:04 +0200
Subject: [PATCH] PONRTSYS-7369: Activate GPTC2 clock in TEP

The TEP watchdog driver uses the GPTC2 clock. When the TEP watchdog is
Initialized it accesses the GPTC2 registers and needs the clock active,
overwise it ends up in a data bus error.

After chip power reset the GPTC2 clock gate is open, the PON SDK 1.7 does
not deactivate this clock, but PON SK 1.6 does.

When someone loads PON SDK 1.6 onto the system first, reboot the system
without a power reset (not POR) and then installs PON SDK 1.7 or later this
clock is still deactivated, because the PON SDK 1.7 does not touch it.
When a power reset (POR) is done in between the CGU gets reset and it is
booting up fine again. A normal SW reset does not reset the CGU registers back
to the reset values.

As we do not have a clock driver in the TEP we directly write to the
CGU_CLKGCR1_A register. This register is only used to activate the clock
gate, it there is a different one to deactivate a clock gate. There is
no manual synchronisation needed.
---
 drivers/clk/intel/clk-prx300.c |  4 ++--
 drivers/watchdog/tep-wdt.c     | 15 +++++++++++++++
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/intel/clk-prx300.c b/drivers/clk/intel/clk-prx300.c
index 8f0a3c0df16a..1c0496731429 100644
--- a/drivers/clk/intel/clk-prx300.c
+++ b/drivers/clk/intel/clk-prx300.c
@@ -172,8 +172,8 @@ static const struct intel_clk_branch prx300_branch_clks[] __initconst = {
 		   12, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_GPTC1, "g_gptc1", NULL, 0, CGU_GATE1,
 		   13, GATE_CLK_HW, 0),
-	INTEL_GATE(PRX300_GCLK_GPTC2, "g_gptc2", NULL, 0, CGU_GATE1,
-		   14, GATE_CLK_HW, 0),
+	INTEL_GATE(PRX300_GCLK_GPTC2, "g_gptc2", NULL, CLK_IGNORE_UNUSED,
+		   CGU_GATE1, 14, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_URT0, "g_uart0", NULL, CLK_IGNORE_UNUSED,
 		   CGU_GATE1, 17, GATE_CLK_HW, 0),
 	INTEL_GATE(PRX300_GCLK_URT1, "g_uart1", NULL, 0, CGU_GATE1,
diff --git a/drivers/watchdog/tep-wdt.c b/drivers/watchdog/tep-wdt.c
index fb2b1b7a5ad3..da0868d186cf 100644
--- a/drivers/watchdog/tep-wdt.c
+++ b/drivers/watchdog/tep-wdt.c
@@ -47,6 +47,10 @@
 #define CON_ONESHOT		BIT(3) /* Stop or continue when overflowing */
 #define CON_EXT			BIT(4) /* 32 or 16 bit */
 
+/* CGU register */
+#define CGU_CLKGCR1_A		0x124
+#define  CGU_CLKGCR1_GPTC2	BIT(14)
+
 enum gptc2 {
 	TIMER_2 = 1,
 	TIMER_3 = 2
@@ -56,6 +60,7 @@ struct tep_wdt {
 	unsigned long clk;
 	struct device *dev;
 	struct regmap *rcu;
+	struct regmap *cgu;
 	void __iomem *gptc;
 	struct watchdog_device *wdd;
 };
@@ -196,6 +201,13 @@ static int tep_watchdog_probe(struct platform_device *pdev)
 		return PTR_ERR(priv->rcu);
 	}
 
+	priv->cgu = syscon_regmap_lookup_by_phandle(pdev->dev.of_node,
+		"intel,wdt-cgu");
+	if (IS_ERR(priv->cgu)) {
+		dev_err(&pdev->dev, "Failed to find intel,wdt-cgu\n");
+		return PTR_ERR(priv->cgu);
+	}
+
 	if (of_address_to_resource(dev->of_node, 0, &res)) {
 		dev_err(&pdev->dev, "Failed to find gptc base\n");
 		return -ENXIO;
@@ -234,6 +246,9 @@ static int tep_watchdog_probe(struct platform_device *pdev)
 	watchdog_set_drvdata(priv->wdd, priv);
 	platform_set_drvdata(pdev, priv);
 
+	/* Enable clk gate for GPTC2 */
+	regmap_write(priv->cgu, CGU_CLKGCR1_A, CGU_CLKGCR1_GPTC2);
+
 	/* Enable WDT reset to RCU for VPEx */
 	regmap_write(priv->rcu, RCU_TEP_WDT_RST, WDT_SCPU_RST_EN);
 
